libraries/Microsoft.Bot.Builder.AI.QnA/Dialogs/QnAMakerDialog.cs

```diff
@@ -505,7 +505,62 @@ protected virtual async Task<QnADialogResponseOptions> GetQnAResponseOptionsAsyn
             };
         }
-        
+
+        /// <summary>
+        /// Displays QnA Result from stepContext through Activity - with first answer from QnA Maker response.
+        /// </summary>
+        /// <param name="stepContext">stepContext.</param>
+        /// <param name="cancellationToken">cancellationToken.</param>
+        /// <returns>An object of Task of type <see cref="DialogTurnResult"></see>.</returns>
+        protected virtual async Task<DialogTurnResult> DisplayQnAResultAsync(WaterfallStepContext stepContext, CancellationToken cancellationToken)
+        {
+            var dialogOptions = ObjectPath.GetPathValue<QnAMakerDialogOptions>(stepContext.ActiveDialog.State, Options);
+            var reply = stepContext.Context.Activity.Text;
+
+            if (reply.Equals(dialogOptions.ResponseOptions.CardNoMatchText, StringComparison.OrdinalIgnoreCase))
+            {
+                var activity = dialogOptions.ResponseOptions.CardNoMatchResponse;
+                if (activity == null)
+                {
+                    await stepContext.Context.SendActivityAsync(DefaultCardNoMatchResponse, cancellationToken: cancellationToken).ConfigureAwait(false);
+                }
+                else
+                {
+                    await stepContext.Context.SendActivityAsync(activity, cancellationToken: cancellationToken).ConfigureAwait(false);
+                }
+
+                return await stepContext.EndDialogAsync().ConfigureAwait(false);
+            }
+
+            // If previous QnAId is present, replace the dialog
+            var previousQnAId = ObjectPath.GetPathValue<int>(stepContext.ActiveDialog.State, PreviousQnAId, 0);
+            if (previousQnAId > 0)
+            {
+                // restart the waterfall to step 0
+                return await RunStepAsync(stepContext, index: 0, reason: DialogReason.BeginCalled, result: null, cancellationToken: cancellationToken).ConfigureAwait(false);
+            }
+
+            // If response is present then show that response, else default answer.
+            if (stepContext.Result is List<QueryResult> response && response.Count > 0)
+            {
+                await stepContext.Context.SendActivityAsync(response.First().Answer, cancellationToken: cancellationToken).ConfigureAwait(false);
+            }
+            else
+            {
+                var activity = dialogOptions.ResponseOptions.NoAnswer;
+                if (activity == null)
+                {
+                    await stepContext.Context.SendActivityAsync(DefaultNoAnswer, cancellationToken: cancellationToken).ConfigureAwait(false);
+                }
+                else
+                {
+                    await stepContext.Context.SendActivityAsync(activity, cancellationToken: cancellationToken).ConfigureAwait(false);
+                }
+            }
+
+            return await stepContext.EndDialogAsync().ConfigureAwait(false);
```

Comment:
stepContext.EndDialogAsync(null, cancellationToken)

Reply:
This change is only on the access modifier and movement because of the method now being protected virtual instead of private. So, would not want to make any other changes to code in this PR.